source/geod24/Queue.d

```diff
@@ -0,0 +1,391 @@
+
+    public bool tryDequeue(out T value)
+    {
+        auto head = this.head;
+        auto tail = this.tail;
+        auto next = head.next;
+
+        if (this.head is head)
+        {
+            if (head is tail)
+            {
+                if (next !is null)
+                {
+                    cas(&this.tail, tail, next);
+                    tail = this.tail;
+                }
+            }
+
+            if (head !is tail)
+            {
+                if (cas(&this.head, head, next))
+                {
+                    value = cast(T)next.value;
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+}
+
+unittest
+{
+    ///  Start `writers` amount of threads to write into a queue.
+    ///  Start `readers` amount of threads to read from the queue.
+    ///  Each writer counts from 0 to `count` and sends each number into the queue.
+    ///  The sum is checked at the end.
+    void test_run(alias Q)(uint writers, uint readers, uint count)
+    {
+        import std.bigint : BigInt;
+        import std.range;
+
+        import core.thread;
+
+        immutable(BigInt) correct_sum = BigInt(count) * BigInt(count-1) / 2 * writers;
+
+        BigInt sum = 0;
+
+        auto q = new Q();
+
+        auto write_worker = ()
+        {
+            Thread[] ts;
+            foreach (i; 0 .. writers)
+            {
+                auto t = new Thread(
+                    {
+                        foreach (n; 1 .. count)
+                            q.enqueue(n);
+                    }
+                );
+                t.start();
+                ts ~= t;
+            }
+
+            foreach (t; ts)
+                t.join();
+        };
+
+        auto read_worker = ()
+        {
+            Thread[] ts;
+            foreach (i; 0 .. readers)
+            {
+                auto t = new Thread(
+                    {
+                        BigInt s = 0;
+                        foreach (_; 1 .. count)
+                        {
+                            auto n = q.dequeue();
+                            s += n;
+                        }
+                        synchronized { sum += s; }
+                    }
+                );
+                t.start();
+                ts ~= t;
+            }
+
+            foreach (t; ts)
+                t.join();
+        };
+
+        auto w = new Thread(write_worker);
+        auto r = new Thread(read_worker);
+
+        w.start();
+        r.start();
+
+        w.join();
+        r.join();
+
+        assert(sum == correct_sum);
+    }
+    enum readers = 10;
+    enum writers = 10;
+    enum count = 10_000;
+
+    void f0 ()
+    {
+        test_run!(BlockingQueue!long)     (writers, readers, count);
+    }
+
+    void f1 ()
+    {
+        test_run!(NonBlockingQueue!long)  (writers, readers, count);
+    }
+
+    import std.datetime.stopwatch : benchmark;
+    auto r = benchmark!(f0, f1)(3);
```

Comment:
Nice!
I recommend the following:
1) Have actual tests for correctness
2) Have a `version (none)` benchmarking code

This way you can easily switch between `version (none)` and `version (all)` during development to benchmark, but you don't slow down / pollute your unittests with benchmark (for example, you don't want to run the benchmark on every CI run, and/or every unit test).